File: apps/web-vite/src/effects/definitions/ai-upscale.ts
import type { EffectDefinition } from "@/effects/types";

export const AI_UPSCALE_SHADER = "ai-upscale";

export const aiUpscaleEffectDefinition: EffectDefinition = {
	type: "ai-upscale",
	name: "AI Upscale",
	keywords: ["upscale", "super-resolution", "enhance", "sharpen", "ai", "anime4k"],
	params: [
		{ key: "intensity", label: "Intensity", type: "number", default: 50, min: 0, max: 100, step: 1 },
	],
	renderer: {
		passes: [{
			shader: AI_UPSCALE_SHADER,
			uniforms: ({ effectParams }) => ({
				u_scalar0: (Number(effectParams.intensity) || 50) / 100,
			}),
		}],
	},
};

File: rust/crates/effects/src/definitions/ai_upscale.rs
use std::collections::HashMap;

use crate::{EffectCategory, EffectDefinition, EffectParam, EffectUniformBuffer, UniformValue};

fn pack_uniforms(
    uniforms: &HashMap<String, UniformValue>,
    width: u32,
    height: u32,
) -> EffectUniformBuffer {
    let s0 = uniforms
        .get("u_scalar0")
        .and_then(|v| match v {
            UniformValue::Number(n) => Some(*n),
            _ => None,
        })
        .unwrap_or(0.0);

    EffectUniformBuffer {
        resolution: [width as f32, height as f32],
        direction: [0.0, 0.0],
        scalars: [s0, 0.0, 0.0, 0.0],
        _cg_expansion: [0u8; 480],
    }
}

pub const AI_UPSCALE: EffectDefinition = EffectDefinition {
    id: "ai-upscale",
    name: "AI Upscale",
    category: EffectCategory::Blur,
    shader_source: include_str!("../shaders/ai_upscale.wgsl"),
    entry_point: "fragment_main",
    uniform_size: std::mem::size_of::<EffectUniformBuffer>(),
    params: &[
        EffectParam {
            name: "u_scalar0",
            label: "Intensity",
            default: 0.0,
            min: 0.0,
            max: 1.0,
            step: 0.01,
            animatable: true,
        },
    ],
    pack_uniforms,
};
